From 69be7f9aacb8d1004cf9023fd9c87d7c97e00dd2 Mon Sep 17 00:00:00 2001
From: Viacheslav Volkov <viacheslav.volkov.1@gmail.com>
Date: Sun, 3 May 2020 17:50:55 +0300
Subject: [PATCH] Add "-selection" option and Alt+i hotkey.

---
 config.mk |  9 ++++-----
 dmenu.1   |  9 ++++-----
 dmenu.c   | 33 ++++++++++++++++++++++++++++-----
 3 files changed, 36 insertions(+), 15 deletions(-)

diff --git a/config.mk b/config.mk
index 0929b4a..2b0766e 100644
--- a/config.mk
+++ b/config.mk
@@ -5,8 +5,7 @@ VERSION = 4.9
 PREFIX = /usr/local
 MANPREFIX = $(PREFIX)/share/man
 
-X11INC = /usr/X11R6/include
-X11LIB = /usr/X11R6/lib
+X11INC = /usr/include
 
 # Xinerama, comment if you don't want it
 XINERAMALIBS  = -lXinerama
@@ -20,12 +19,12 @@ FREETYPEINC = /usr/include/freetype2
 
 # includes and libs
 INCS = -I$(X11INC) -I$(FREETYPEINC)
-LIBS = -L$(X11LIB) -lX11 $(XINERAMALIBS) $(FREETYPELIBS)
+LIBS = -lX11 $(XINERAMALIBS) $(FREETYPELIBS) -lXmu
 
 # flags
-CPPFLAGS = -D_DEFAULT_SOURCE -D_BSD_SOURCE -D_XOPEN_SOURCE=700 -D_POSIX_C_SOURCE=200809L -DVERSION=\"$(VERSION)\" $(XINERAMAFLAGS)
+CPPFLAGS = -D_DEFAULT_SOURCE -D_BSD_SOURCE -D_XOPEN_SOURCE=700 -D_POSIX_C_SOURCE=200809L -DVERSION=\"$(VERSION)\" $(XINERAMAFLAGS) -O2 -flto
 CFLAGS   = -std=c99 -pedantic -Wall -Os $(INCS) $(CPPFLAGS)
-LDFLAGS  = $(LIBS)
+LDFLAGS  = $(LIBS) -O2 -flto
 
 # compiler and linker
 CC = cc
diff --git a/dmenu.1 b/dmenu.1
index 323f93c..edd66fb 100644
--- a/dmenu.1
+++ b/dmenu.1
@@ -10,6 +10,8 @@ dmenu \- dynamic menu
 .IR monitor ]
 .RB [ \-p
 .IR prompt ]
+.RB [ \-selection
+.IR {primary | secondary | clipboard} ]
 .RB [ \-fn
 .IR font ]
 .RB [ \-nb
@@ -160,11 +162,8 @@ Delete line left
 .B C\-w
 Delete word left
 .TP
-.B C\-y
-Paste from primary X selection
-.TP
-.B C\-Y
-Paste from X clipboard
+.B A\-v
+Paste from clipboard
 .TP
 .B M\-b
 Move cursor to the start of the current word
diff --git a/dmenu.c b/dmenu.c
index 6b8f51b..bbf7c7c 100644
--- a/dmenu.c
+++ b/dmenu.c
@@ -11,6 +11,7 @@
 #include <X11/Xlib.h>
 #include <X11/Xatom.h>
 #include <X11/Xutil.h>
+#include <X11/Xmu/Atoms.h>
 #ifdef XINERAMA
 #include <X11/extensions/Xinerama.h>
 #endif
@@ -28,6 +29,13 @@
 /* enums */
 enum { SchemeNorm, SchemeSel, SchemeOut, SchemeLast }; /* color schemes */
 
+enum Clipboard
+{
+	PRIMARY,
+	SECONDARY,
+	CLIPBOARD,
+} clipboard = CLIPBOARD;
+
 struct item {
 	char *text;
 	struct item *left, *right;
@@ -354,11 +362,6 @@ keypress(XKeyEvent *ev)
 			while (cursor > 0 && !strchr(worddelimiters, text[nextrune(-1)]))
 				insert(NULL, nextrune(-1) - cursor);
 			break;
-		case XK_y: /* paste selection */
-		case XK_Y:
-			XConvertSelection(dpy, (ev->state & ShiftMask) ? clip : XA_PRIMARY,
-			                  utf8, utf8, win, CurrentTime);
-			return;
 		case XK_Left:
 			movewordedge(-1);
 			goto draw;
@@ -388,6 +391,18 @@ keypress(XKeyEvent *ev)
 		case XK_j: ksym = XK_Next;  break;
 		case XK_k: ksym = XK_Prior; break;
 		case XK_l: ksym = XK_Down;  break;
+		case XK_v: // paste (insert) from clipboard
+		{
+			Atom sel;
+			switch (clipboard)
+			{
+				case PRIMARY: sel = XA_PRIMARY; break;
+				case SECONDARY: sel = XA_SECONDARY; break;
+				case CLIPBOARD: sel = XA_CLIPBOARD(dpy); break;
+			}
+			XConvertSelection(dpy, sel, utf8, utf8, win, CurrentTime);
+			return;
+		}
 		default:
 			return;
 		}
@@ -698,6 +713,14 @@ main(int argc, char *argv[])
 		if (!strcmp(argv[i], "-v")) {      /* prints version information */
 			puts("dmenu-"VERSION);
 			exit(0);
+		} else if (!strcmp(argv[i], "-selection")) {
+			++i;
+			if (!strcmp(argv[i], "primary"))
+				clipboard = PRIMARY;
+			else if (!strcmp(argv[i], "secondary"))
+				clipboard = SECONDARY;
+			else
+				clipboard = CLIPBOARD;
 		} else if (!strcmp(argv[i], "-b")) /* appears at the bottom of the screen */
 			topbar = 0;
 		else if (!strcmp(argv[i], "-f"))   /* grabs keyboard before reading stdin */
-- 
2.26.2

