diff --git a/config.def.h b/config.def.h
index 0e01717..5d77ac7 100644
--- a/config.def.h
+++ b/config.def.h
@@ -5,8 +5,8 @@
  *
  * font: see http://freedesktop.org/software/fontconfig/fontconfig-user.html
  */
-static char *font = "Liberation Mono:pixelsize=12:antialias=true:autohint=true";
-static int borderpx = 2;
+static char *font = "Inconsolata LGC:pixelsize=17";
+static int borderpx = 0;
 
 /*
  * What program is execed by st depends of these precedence rules:
@@ -84,31 +84,23 @@ unsigned int tabspaces = 8;
 
 /* Terminal colors (16 first used in escape sequence) */
 static const char *colorname[] = {
-	/* 8 normal colors */
-	"black",
-	"red3",
-	"green3",
-	"yellow3",
-	"blue2",
-	"magenta3",
-	"cyan3",
-	"gray90",
-
-	/* 8 bright colors */
-	"gray50",
-	"red",
-	"green",
-	"yellow",
-	"#5c5cff",
-	"magenta",
-	"cyan",
-	"white",
-
-	[255] = 0,
-
-	/* more colors can be added after 255 to use with DefaultXX */
-	"#cccccc",
-	"#555555",
+	/* Molokai theme: */
+	"#101010",
+	"#960050",
+	"#66aa11",
+	"#c47f2c",
+	"#30309b",
+	"#7e40a5",
+	"#3579a8",
+	"#9999aa",
+	"#303030",
+	"#ff0090",
+	"#80ff00",
+	"#ffba68",
+	"#5f5fee",
+	"#bb88dd",
+	"#4eb4fa",
+	"#d0d0d0",
 };
 
 
@@ -116,9 +108,9 @@ static const char *colorname[] = {
  * Default colors (colorname index)
  * foreground, background, cursor, reverse cursor
  */
-unsigned int defaultfg = 7;
+unsigned int defaultfg = 15;
 unsigned int defaultbg = 0;
-static unsigned int defaultcs = 256;
+static unsigned int defaultcs = 15;
 static unsigned int defaultrcs = 257;
 
 /*
