From 201ba54d86b63e045346f283576b856186f299fe Mon Sep 17 00:00:00 2001
From: Viacheslav Volkov <viacheslav.volkov.1@gmail.com>
Date: Sun, 16 Oct 2016 19:06:55 +0200
Subject: [PATCH] Add "-selection" option and Alt+i hotkey.

---
 config.mk |  9 ++++-----
 dmenu.1   |  9 ++++-----
 dmenu.c   | 33 ++++++++++++++++++++++++++++-----
 3 files changed, 36 insertions(+), 15 deletions(-)

diff --git a/config.mk b/config.mk
index 4d908a5..955a0bc 100644
--- a/config.mk
+++ b/config.mk
@@ -5,8 +5,7 @@ VERSION = 4.6
 PREFIX = /usr/local
 MANPREFIX = ${PREFIX}/share/man
 
-X11INC = /usr/X11R6/include
-X11LIB = /usr/X11R6/lib
+X11INC = /usr/include
 
 # Xinerama, comment if you don't want it
 XINERAMALIBS  = -lXinerama
@@ -20,12 +19,12 @@ FREETYPEINC = /usr/include/freetype2
 
 # includes and libs
 INCS = -I${X11INC} -I${FREETYPEINC}
-LIBS = -L${X11LIB} -lX11 ${XINERAMALIBS} ${FREETYPELIBS}
+LIBS = -lX11 ${XINERAMALIBS} ${FREETYPELIBS} -lXmu
 
 # flags
-CPPFLAGS = -D_DEFAULT_SOURCE -D_BSD_SOURCE -D_XOPEN_SOURCE=700 -D_POSIX_C_SOURCE=200809L -DVERSION=\"${VERSION}\" ${XINERAMAFLAGS}
+CPPFLAGS = -D_DEFAULT_SOURCE -D_BSD_SOURCE -D_XOPEN_SOURCE=700 -D_POSIX_C_SOURCE=200809L -DVERSION=\"${VERSION}\" ${XINERAMAFLAGS} -O2 -flto
 CFLAGS   = -std=c99 -pedantic -Wall -Os ${INCS} ${CPPFLAGS}
-LDFLAGS  = -s ${LIBS}
+LDFLAGS  = -s ${LIBS} -O2 -flto
 
 # compiler and linker
 CC = cc
diff --git a/dmenu.1 b/dmenu.1
index 9eab758..cd61b64 100644
--- a/dmenu.1
+++ b/dmenu.1
@@ -10,6 +10,8 @@ dmenu \- dynamic menu
 .IR monitor ]
 .RB [ \-p
 .IR prompt ]
+.RB [ \-selection
+.IR {primary | secondary | clipboard} ]
 .RB [ \-fn
 .IR font ]
 .RB [ \-nb
@@ -154,11 +156,8 @@ Delete line left
 C\-w
 Delete word left
 .TP
-C\-y
-Paste from primary X selection
-.TP
-C\-Y
-Paste from X clipboard
+A\-i
+Paste from clipboard
 .TP
 M\-g
 Home
diff --git a/dmenu.c b/dmenu.c
index 9278e91..43c67d8 100644
--- a/dmenu.c
+++ b/dmenu.c
@@ -10,6 +10,7 @@
 #include <X11/Xlib.h>
 #include <X11/Xatom.h>
 #include <X11/Xutil.h>
+#include <X11/Xmu/Atoms.h>
 #ifdef XINERAMA
 #include <X11/extensions/Xinerama.h>
 #endif
@@ -27,6 +28,13 @@
 /* enums */
 enum { SchemeNorm, SchemeSel, SchemeOut, SchemeLast }; /* color schemes */
 
+enum  Clipboard
+{
+	PRIMARY,
+	SECONDARY,
+	CLIPBOARD,
+} clipboard = CLIPBOARD;
+
 struct item {
 	char *text;
 	struct item *left, *right;
@@ -329,11 +337,6 @@ keypress(XKeyEvent *ev)
 			while (cursor > 0 && !strchr(worddelimiters, text[nextrune(-1)]))
 				insert(NULL, nextrune(-1) - cursor);
 			break;
-		case XK_y: /* paste selection */
-		case XK_Y:
-			XConvertSelection(dpy, (ev->state & ShiftMask) ? clip : XA_PRIMARY,
-			                  utf8, utf8, win, CurrentTime);
-			return;
 		case XK_Return:
 		case XK_KP_Enter:
 			break;
@@ -351,6 +354,18 @@ keypress(XKeyEvent *ev)
 		case XK_j: ksym = XK_Next;  break;
 		case XK_k: ksym = XK_Prior; break;
 		case XK_l: ksym = XK_Down;  break;
+		case XK_i: // paste (insert) from clipboard
+		{
+			Atom sel;
+			switch (clipboard)
+			{
+				case PRIMARY: sel = XA_PRIMARY; break;
+				case SECONDARY: sel = XA_SECONDARY; break;
+				case CLIPBOARD: sel = XA_CLIPBOARD(dpy); break;
+			}
+			XConvertSelection(dpy, sel, utf8, utf8, win, CurrentTime);
+			return;
+		}
 		default:
 			return;
 		}
@@ -651,6 +666,14 @@ main(int argc, char *argv[])
 		if (!strcmp(argv[i], "-v")) {      /* prints version information */
 			puts("dmenu-"VERSION);
 			exit(0);
+		} else if (!strcmp(argv[i], "-selection")) {
+			++i;
+			if(!strcmp(argv[i], "primary"))
+				clipboard = PRIMARY;
+			else if (!strcmp(argv[i], "secondary"))
+				clipboard = SECONDARY;
+			else
+				clipboard = CLIPBOARD;
 		} else if (!strcmp(argv[i], "-b")) /* appears at the bottom of the screen */
 			topbar = 0;
 		else if (!strcmp(argv[i], "-f"))   /* grabs keyboard before reading stdin */
-- 
2.1.4

