diff -urp a/config.mk b/config.mk
--- a/config.mk
+++ b/config.mk
@@ -18,13 +18,13 @@ XFTLIBS  = -lXft -lXrender -lfreetype -l
 
 # includes and libs
 INCS = -I${X11INC} ${XFTINC}
-LIBS = -L${X11LIB} -lX11 ${XINERAMALIBS} ${XFTLIBS}
+LIBS = -L${X11LIB} -lX11 ${XINERAMALIBS} ${XFTLIBS} -lXmu
 
 # flags
 CPPFLAGS = -D_BSD_SOURCE -D_POSIX_C_SOURCE=2 -DVERSION=\"${VERSION}\" ${XINERAMAFLAGS}
 #CFLAGS   = -g -std=c99 -pedantic -Wall -O0 ${INCS} ${CPPFLAGS}
-CFLAGS   = -std=c99 -pedantic -Wall -Os ${INCS} ${CPPFLAGS}
-LDFLAGS  = -s ${LIBS}
+CFLAGS   = -std=c99 -pedantic -Wall ${INCS} ${CPPFLAGS} -O2 -flto
+LDFLAGS  = -s ${LIBS} -O2 -flto
 
 # compiler and linker
 CC = cc
diff -urp a/dmenu.1 b/dmenu.1
--- a/dmenu.1
+++ b/dmenu.1
@@ -10,6 +10,8 @@ dmenu \- dynamic menu
 .IR lines ]
 .RB [ \-p
 .IR prompt ]
+.RB [ \-selection
+.IR {primary | secondary | clipboard} ]
 .RB [ \-fn
 .IR font ]
 .RB [ \-nb
@@ -137,11 +139,8 @@ Delete line left
 C\-w
 Delete word left
 .TP
-C\-y
-Paste from primary X selection
-.TP
-C\-Y
-Paste from X clipboard
+A\-i
+Paste from clipboard
 .TP
 M\-g
 Home
diff -urp a/dmenu.c b/dmenu.c
--- a/dmenu.c
+++ b/dmenu.c
@@ -8,6 +8,7 @@
 #include <X11/Xlib.h>
 #include <X11/Xatom.h>
 #include <X11/Xutil.h>
+#include <X11/Xmu/Atoms.h>
 #ifdef XINERAMA
 #include <X11/extensions/Xinerama.h>
 #endif
@@ -68,6 +69,13 @@ static XIC xic;
 static int (*fstrncmp)(const char *, const char *, size_t) = strncmp;
 static char *(*fstrstr)(const char *, const char *) = strstr;
 
+enum CLIPBOARD
+{
+	PRIMARY,
+	SECONDARY,
+	CLIPBOARD
+} clipboard = CLIPBOARD;
+
 int
 main(int argc, char *argv[]) {
 	Bool fast = False;
@@ -90,6 +98,16 @@ main(int argc, char *argv[]) {
 		else if(i+1 == argc)
 			usage();
 		/* these options take one argument */
+		else if(!strcmp(argv[i], "-selection"))
+		{
+			++i;
+			if(!strcmp(argv[i], "primary"))
+				clipboard = PRIMARY;
+			else if(!strcmp(argv[i], "secondary"))
+				clipboard = SECONDARY;
+			else
+				clipboard = CLIPBOARD;
+		}
 		else if(!strcmp(argv[i], "-l"))   /* number of lines in vertical list */
 			lines = atoi(argv[++i]);
 		else if(!strcmp(argv[i], "-p"))   /* adds prompt to left of input field */
@@ -287,10 +305,6 @@ keypress(XKeyEvent *ev) {
 			while(cursor > 0 && text[nextrune(-1)] != ' ')
 				insert(NULL, nextrune(-1) - cursor);
 			break;
-		case XK_y: /* paste selection */
-			XConvertSelection(dc->dpy, (ev->state & ShiftMask) ? clip : XA_PRIMARY,
-			                  utf8, utf8, win, CurrentTime);
-			return;
 		default:
 			return;
 		}
@@ -302,6 +316,18 @@ keypress(XKeyEvent *ev) {
 		case XK_j: ksym = XK_Next;  break;
 		case XK_k: ksym = XK_Prior; break;
 		case XK_l: ksym = XK_Down;  break;
+		case XK_i: // paste (insert) from clipboard
+		{
+			Atom sel;
+			switch (clipboard)
+			{
+				case PRIMARY: sel = XA_PRIMARY; break;
+				case SECONDARY: sel = XA_SECONDARY; break;
+				case CLIPBOARD: sel = XA_CLIPBOARD(dc->dpy); break;
+			}
+			XConvertSelection(dc->dpy, sel, utf8, utf8, win, CurrentTime);
+			return;
+		}
 		default:
 			return;
 		}
